### ISSUE TITLE
builddependencies not loaded on the extensions step

### ISSUE THREAD

**Issue by damianam** [hpcugent/easybuild-framework#1741]
As discussed in our last telco: `Perl `with `GCC 5.3.0` and `binutils 2.26` doesn't work. The linking of an extension failed, because the linker used was `/usr/bin/ld.gold` instead of the linker present in `binutils 2.26`. `binutils` was declared as a build dependency. When changing that to a normal dependency everything works as expected. I suspect that the `builddependencies` are not loaded in the extensions step. Attached is the easyconfig used. However, it doesn't use an official toolchain, so small tuning might be required to reproduce this. In principle I guess it should be reproducible in `GCC[core] 4.9.3` based toolchains, as long as `binutils `is` 2.26`. Additionally, this would passed inadvertently if the system linker was recent enough.

[Perl-5.22.1-GCCcore-5.3.0.eb.txt](https://github.com/hpcugent/easybuild-framework/files/238947/Perl-5.22.1-GCCcore-5.3.0.eb.txt)

**Comment by boegel** [comment]
@damianam just took a quick look at the code, I think you're right...

basically, what happens when installing extensions is that the temporary module that is created during the installation process is loaded after resetting the environment to how it was before the build started, to prepare the environment for installing the extensions... (see the call to `load_fake_module` in `extensions_step`)

and since no `load` statements for the build depenencies are included in the generated module file...

**Comment by boegel** [comment]
On the other hand, the `toolchain.prepare` method is called every time a new extension in 'created' (i.e. when a new `Extension` instance is created), which should result in (re)loading all dependencies (incl. build deps)...

I'll see if I can reproduce this first.

@damianam if you could share a debug log, that would be very useful.

**Comment by damianam** [comment]
There you go

[easybuild-Perl-5.22.1-20160428.091751.qUgXd.log.gz](https://github.com/hpcugent/easybuild-framework/files/240056/easybuild-Perl-5.22.1-20160428.091751.qUgXd.log.gz)

**Comment by boegel** [comment]
OK, I've figured out the problem...

When an `Extension` instance is created, the `EasyConfig` instance from the parent is copied, but without making sure that the `Toolchain` instance inside of it is properly copied as well. This causes the list of (build) dependencies that were added to disappear, causing the corresponding modules not to be loaded for extensions.
For regular dependencies, that's fine, because these get loaded via the loading of the fake module already at the start of the `extensions` step.

I'll look into a good way of fixing this, there are a couple of options here...
